datafusion/src/physical_plan/sorts/external_sort.rs

```diff
@@ -0,0 +1,696 @@
+    metrics: ExecutionPlanMetricsSet,
+    used: AtomicIsize,
+    spilled_bytes: AtomicUsize,
+    spilled_count: AtomicUsize,
+    insert_finished: AtomicBool,
+}
+
+impl ExternalSorter {
+    pub fn new(
+        partition_id: usize,
+        schema: SchemaRef,
+        expr: Vec<PhysicalSortExpr>,
+        runtime: Arc<RuntimeEnv>,
+    ) -> Self {
+        Self {
+            id: MemoryConsumerId::new(partition_id),
+            schema,
+            in_mem_batches: Mutex::new(vec![]),
+            spills: Mutex::new(vec![]),
+            expr,
+            runtime,
+            metrics: ExecutionPlanMetricsSet::new(),
+            used: AtomicIsize::new(0),
+            spilled_bytes: AtomicUsize::new(0),
+            spilled_count: AtomicUsize::new(0),
+            insert_finished: AtomicBool::new(false),
+        }
+    }
+
+    pub(crate) fn finish_insert(&self) {
+        self.insert_finished.store(true, Ordering::SeqCst);
+    }
+
+    async fn spill_while_inserting(&self) -> Result<usize> {
+        info!(
+            "{} spilling sort data of {} to disk while inserting ({} time(s) so far)",
+            self.str_repr(),
+            self.get_used(),
+            self.spilled_count()
+        );
+
+        let partition = self.partition_id();
+        let mut in_mem_batches = self.in_mem_batches.lock().await;
+        // we could always get a chance to free some memory as long as we are holding some
+        if in_mem_batches.len() == 0 {
+            return Ok(0);
+        }
+
+        let baseline_metrics = BaselineMetrics::new(&self.metrics, partition);
+
+        let path = self.runtime.disk_manager.create_tmp_file()?;
+        let stream = in_mem_merge_sort(
+            &mut *in_mem_batches,
+            self.schema.clone(),
+            &*self.expr,
+            self.runtime.batch_size(),
+            baseline_metrics,
+        )
+        .await;
+
+        let total_size = spill(&mut stream?, path.clone(), self.schema.clone()).await?;
+
+        let mut spills = self.spills.lock().await;
+        self.spilled_count.fetch_add(1, Ordering::SeqCst);
+        self.spilled_bytes.fetch_add(total_size, Ordering::SeqCst);
+        spills.push(path);
+        Ok(total_size)
+    }
+
+    async fn insert_batch(&self, input: RecordBatch) -> Result<()> {
+        let size = batch_memory_size(&input);
+        self.allocate(size).await?;
+        // sort each batch as it's inserted, more probably to be cache-resident
+        let sorted_batch = sort_batch(input, self.schema.clone(), &*self.expr)?;
+        let mut in_mem_batches = self.in_mem_batches.lock().await;
+        in_mem_batches.push(sorted_batch);
+        Ok(())
+    }
+
+    /// MergeSort in mem batches as well as spills into total order with `SortPreservingMergeStream`(SPMS).
+    /// Always put in mem batch based stream to idx 0 in SPMS so that we could spill
+    /// the stream when `spill()` is called on us.
+    async fn sort(&self) -> Result<SendableRecordBatchStream> {
```

Comment:
The external sort here is a little bit different here, it's acquiring memory budget at each record batch basis, instead of getting a huge budget of fixed size ahead of real usage. By acquiring memory for record batch at a time, we can get finer control of how much memory is really used in the system. Therefore, the external sort implementation here is:

1. Get one record batch from input, acquire a memory budget of its size from the memory manager, sort the batch, and insert the batch to a vector that contains all in-memory, buffered batches.
2. when the external sorter is required to spill to disk, (triggered by other consumers or just itself when no memory available) it first looks up if itself contains in-memory buffered batches, if so, it does a heap sort to merge all in-memory, self-sorted batches to a partial order, and writes to file, frees up all memory the vector contains.
3. when all input batches are seen, do a merge sort of in-memory batches with all spills, resulting in a final, total order.

Reply:
Makes sense @yjshen  -- I think there is a tradeoff here:

 If memory is allocated "on demand" for each record batch as you describe in this ticket, one operator can effectively consume all the memory, causing others to spill even when they might not need to. However, the on demand strategy does work well to ensure that all memory allocated is used

For example, if you have a plan like

```
Sort
  GroupByHash
```

If you allocate memory on demand, it is possible that the group by hash consumes most/all of the memory available (and still requires spilling) so by the time the sort is run, it can't get much/all memory and so it spills unecessairly

However, it is true that if memory is allocated up front, there is the very real risk of allocating more memory to sort or group by hash than can be used (resulting in waste)

in conclusion, I think the algorithm you describe sounds like a good one

Reply:
For the Sort -> GroupBy -> ParquetScan case, and for most operator cases, the memory usage style is clear: a growing phase and a shrinking phase. 

Give group by as an example; it would use up all available memory and require spills during hash building. i.e., the growing phase. When all inputs are inserted, it enters into a shrinking mode, while it may still hold some memory for output, but also spillable to first write its aggregation result to a file and then read file by its successor. Memory usage can only reduce since then.

However, bad guys that output while buffering while inputting, e.g., a window operator, do not behave well like grow-then-shrink. And if we have several windows chained in a single stage, the spilling behavior would likely be strange, often spills for a single operator, and hard to proceed for the entire stage. (For Spark's memory management, as well for my current implementation).

Two available solutions for this case are:
1. allocation strategist priority aware. i.e., while insufficient memory is witnessed, we only trigger consumers of a lower priority to spill and throttle processing speed by starving the childing operators.
2. executor aware scheduler, as brought out several times in the community, throttle records/batch processing speed for different operators by scheduling them with adjustable speed.

Reply:
> Give group by as an example; it would use up all available memory and require spills during hash building. i.e., the growing phase. When all inputs are inserted, it enters into a shrinking mode, while it may still hold some memory for output, but also spillable to first write its aggregation result to a file and then read file by its successor. Memory usage can only reduce since then.

The classic group by spill algorithm I am familiar with  spills groups to disk, sorted by group key when the hash table memory is exhausted.

Then, data is output by merging the sorted runs from disk.

upon thinking about this, it does seem reasonable that the merge phase can end up using much less memory 👍 

> However, bad guys that output while buffering while inputting, e.g., a window operator, do not behave well like grow-then-shrink.

That is a good point